File: packages/knowledge-units/python/intelliengine_knowledge_units/runtime.py
from __future__ import annotations

import copy
import re
from pathlib import Path
from typing import Any

from intelliengine_conformance.json_codec import JsonInputError, parse_json_bytes
from intelliengine_conformance.schema_validation import is_valid


JsonObject = dict[str, Any]
SAFE_INTEGER = 9007199254740991
UUID = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")
CAPABILITIES = {
    "runtime.math.numeric",
    "runtime.math.symbolic",
    "runtime.visualization.2d",
}
REQUIRED = [
    "contract_version", "id", "revision", "title", "concept_boundary",
    "learning_objectives", "node_bindings", "prerequisite_unit_refs",
    "behaviors", "validations", "mastery_criteria", "provenance_refs",
]


def _issue(code: str, path: str) -> JsonObject:
    return {"code": code, "path": path, "severity": "error"}


def _result(valid: bool, issue: JsonObject | None = None) -> JsonObject:
    return {
        "object_result": "valid" if valid else "invalid",
        "operation_outcome": "succeeded",
        "issues": [] if issue is None else [issue],
    }


def _ref_key(value: Any) -> tuple[bytes, int] | None:
    if not isinstance(value, dict) or set(value) != {"id", "revision"}:
        return None
    identifier = value.get("id")
    revision = value.get("revision")
    if not isinstance(identifier, str) or UUID.fullmatch(identifier) is None or isinstance(revision, bool) or not isinstance(revision, int):
        return None
    if not 1 <= revision <= SAFE_INTEGER:
        return None
    return identifier.encode("utf-8"), revision


def _sorted_unique(values: Any, key) -> bool:
    if not isinstance(values, list):
        return False
    keys = [key(value) for value in values]
    return all(item is not None for item in keys) and len(keys) == len(set(keys)) and keys == sorted(keys)


def _binding_key(value: Any):
    if not isinstance(value, dict):
        return None
    role = value.get("role")
    ref = _ref_key(value.get("node_ref"))
    return None if not isinstance(role, str) or ref is None else (role.encode("utf-8"), *ref)


def _named_key(field: str):
    return lambda value: value[field].encode("utf-8") if isinstance(value, dict) and isinstance(value.get(field), str) else None


def _collect_refs(unit: JsonObject) -> list[tuple[bytes, int]]:
    values: list[Any] = []
    boundary = unit.get("concept_boundary", {})
    if isinstance(boundary, dict):
        values.extend(boundary.get("focus_node_refs", []))
    for objective in unit.get("learning_objectives", []):
        if isinstance(objective, dict):
            values.extend(objective.get("target_node_refs", []))
    for binding in unit.get("node_bindings", []):
        if isinstance(binding, dict):
            values.append(binding.get("node_ref"))
    for behavior in unit.get("behaviors", []):
        if isinstance(behavior, dict):
            values.extend(behavior.get("input_node_refs", []))
            values.extend(behavior.get("output_node_refs", []))
    for validation in unit.get("validations", []):
        if isinstance(validation, dict):
            values.extend(validation.get("subject_node_refs", []))
            values.extend(validation.get("evidence_node_refs", []))
    for criterion in unit.get("mastery_criteria", []):
        if isinstance(criterion, dict):
            values.extend(criterion.get("evidence_node_refs", []))
    return [key for value in values if (key := _ref_key(value)) is not None]


def validate_unit(unit: Any, available_node_refs: Any, contract_root: Path) -> JsonObject:
    if not isinstance(unit, dict):
        return _result(False, _issue("knowledge_unit.invalid_json", ""))
    missing = next((field for field in REQUIRED if field not in unit), None)
    if missing is not None:
        return _result(False, _issue("knowledge_unit.missing_field", f"/{missing}"))
    if not isinstance(unit["contract_version"], str) or not unit["contract_version"].startswith("1."):
        return _result(False, _issue("knowledge_unit.unsupported_contract_version", "/contract_version"))
    if not isinstance(unit["id"], str) or UUID.fullmatch(unit["id"]) is None:
        return _result(False, _issue("knowledge_unit.invalid_id", "/id"))
    if isinstance(unit["revision"], bool) or not isinstance(unit["revision"], int) or not 1 <= unit["revision"] <= SAFE_INTEGER:
        return _result(False, _issue("knowledge_unit.invalid_revision", "/revision"))
    if not _sorted_unique(unit["node_bindings"], _binding_key):
        return _result(False, _issue("knowledge_unit.noncanonical_set", "/node_bindings"))
    if not _sorted_unique(unit["prerequisite_unit_refs"], _ref_key):
        return _result(False, _issue("knowledge_unit.noncanonical_set", "/prerequisite_unit_refs"))
    if not _sorted_unique(unit["provenance_refs"], lambda value: value.encode("utf-8") if isinstance(value, str) and value else None):
        return _result(False, _issue("knowledge_unit.noncanonical_set", "/provenance_refs"))
    for field, name in (("learning_objectives", "objective_id"), ("behaviors", "behavior_id"), ("validations", "validation_id"), ("mastery_criteria", "criterion_id")):
        if not _sorted_unique(unit[field], _named_key(name)):
            return _result(False, _issue("knowledge_unit.noncanonical_set", f"/{field}"))
    identity = (unit["id"].encode("utf-8"), unit["revision"])
    for index, reference in enumerate(unit["prerequisite_unit_refs"]):
        if _ref_key(reference) == identity:
            return _result(False, _issue("knowledge_unit.self_dependency", f"/prerequisite_unit_refs/{index}"))
    for index, criterion in enumerate(unit["mastery_criteria"]):
        if not isinstance(criterion, dict) or not criterion.get("evidence_node_refs"):
            return _result(False, _issue("knowledge_unit.mastery_without_evidence", f"/mastery_criteria/{index}/evidence_node_refs"))
    for index, behavior in enumerate(unit["behaviors"]):
        if not isinstance(behavior, dict) or behavior.get("capability") not in CAPABILITIES:
            return _result(False, _issue("knowledge_unit.invalid_behavior_capability", f"/behaviors/{index}/capability"))
    if not _sorted_unique(available_node_refs, _ref_key):
        return _result(False, _issue("knowledge_unit.noncanonical_set", "/available_node_refs"))
    available = {_ref_key(reference) for reference in available_node_refs}
    refs = _collect_refs(unit)
    if any(reference not in available for reference in refs):
        return _result(False, _issue("knowledge_unit.dangling_node_ref", ""))
    bound = {_ref_key(binding["node_ref"]) for binding in unit["node_bindings"]}
    if any(reference not in bound for reference in refs):
        return _result(False, _issue("knowledge_unit.dangling_node_ref", "/node_bindings"))
    nested_sets: list[tuple[str, Any]] = []
    boundary = unit.get("concept_boundary")
    if isinstance(boundary, dict):
        nested_sets.append(("/concept_boundary/focus_node_refs", boundary.get("focus_node_refs")))
        if not _sorted_unique(boundary.get("out_of_scope_statements"), lambda value: value.encode("utf-8") if isinstance(value, str) and value else None):
            return _result(False, _issue("knowledge_unit.noncanonical_set", "/concept_boundary/out_of_scope_statements"))
    for index, objective in enumerate(unit["learning_objectives"]):
        nested_sets.append((f"/learning_objectives/{index}/target_node_refs", objective.get("target_node_refs")))
    for index, behavior in enumerate(unit["behaviors"]):
        nested_sets.append((f"/behaviors/{index}/input_node_refs", behavior.get("input_node_refs")))
        nested_sets.append((f"/behaviors/{index}/output_node_refs", behavior.get("output_node_refs")))
    for index, validation in enumerate(unit["validations"]):
        nested_sets.append((f"/validations/{index}/subject_node_refs", validation.get("subject_node_refs")))
        nested_sets.append((f"/validations/{index}/evidence_node_refs", validation.get("evidence_node_refs")))
    for index, criterion in enumerate(unit["mastery_criteria"]):
        nested_sets.append((f"/mastery_criteria/{index}/evidence_node_refs", criterion.get("evidence_node_refs")))
    for path, references in nested_sets:
        if not _sorted_unique(references, _ref_key):
            return _result(False, _issue("knowledge_unit.noncanonical_set", path))
    schema = parse_json_bytes((contract_root / "schemas" / "knowledge-unit.schema.json").read_bytes())
    if not is_valid(unit, schema, schema):
        return _result(False, _issue("knowledge_unit.invalid_json", ""))
    return _result(True)


def parse_and_validate(raw: bytes, available_node_refs: Any, contract_root: Path) -> JsonObject:
    try:
        value = parse_json_bytes(raw)
    except JsonInputError:
        return _result(False, _issue("knowledge_unit.invalid_json", ""))
    return validate_unit(value, available_node_refs, contract_root)


def _walk(document: Any, pointer: str) -> tuple[Any, str | int]:
    parts = [part.replace("~1", "/").replace("~0", "~") for part in pointer[1:].split("/")]
    current = document
    for part in parts[:-1]:
        current = current[int(part)] if isinstance(current, list) else current[part]
    return current, int(parts[-1]) if isinstance(current, list) else parts[-1]


def _mutate(unit: JsonObject, mutation: JsonObject) -> JsonObject:
    value = copy.deepcopy(unit)
    parent, leaf = _walk(value, mutation["path"])
    kind = mutation["kind"]
    if kind == "remove":
        del parent[leaf]
    elif kind == "reverse":
        parent[leaf] = list(reversed(parent[leaf]))
    elif kind == "clear":
        parent[leaf] = []
    elif kind == "replace":
        parent[leaf] = copy.deepcopy(mutation["value"])
    elif kind == "append-self-dependency":
        parent[leaf].append({"id": value["id"], "revision": value["revision"]})
    else:
        raise ValueError(f"unsupported fixture mutation: {kind}")
    return value


def run_fixture_suite(contract_root: Path) -> list[JsonObject]:
    suite = parse_json_bytes((contract_root / "fixtures" / "cases.json").read_bytes())
    base = next(case for case in suite["cases"] if "unit" in case["input"])
    base_unit = base["input"]["unit"]
    available = base["input"]["available_node_refs"]
    rows = []
    for case in suite["cases"]:
        unit = copy.deepcopy(case["input"]["unit"]) if "unit" in case["input"] else _mutate(base_unit, case["input"]["mutation"])
        rows.append({"case_id": case["case_id"], **validate_unit(unit, available, contract_root)})
    return sorted(rows, key=lambda row: row["case_id"].encode("utf-8"))

File: packages/knowledge-units/src/knowledge-unit/runtime.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

import { validateMachineSchema } from "../../../cognitive-ir/src/conformance-ts/machine-schema.ts";
import { StrictJsonError, canonicalize, strictParse } from "../../../cognitive-ir/src/conformance-ts/strict-json.ts";


type JsonObject = Record<string, any>;
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const KNOWLEDGE_UNIT_JCS_BYTES = 1_048_576;
const CAPABILITIES = new Set([
  "runtime.math.numeric",
  "runtime.math.symbolic",
  "runtime.visualization.2d",
]);
const REQUIRED = [
  "contract_version", "id", "revision", "title", "concept_boundary",
  "learning_objectives", "node_bindings", "prerequisite_unit_refs",
  "behaviors", "validations", "mastery_criteria", "provenance_refs",
];


function object(value: unknown): value is JsonObject {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}


function issue(code: string, path: string) {
  return { code, path, severity: "error" };
}


function result(valid: boolean, problem?: JsonObject) {
  return {
    object_result: valid ? "valid" : "invalid",
    operation_outcome: "succeeded",
    issues: problem === undefined ? [] : [problem],
  };
}


function resourceExhausted() {
  return {
    object_result: "not_evaluated",
    operation_outcome: "resource_exhausted",
    issues: [issue("knowledge_unit.invalid_json", "")],
  };
}


function compareBytes(left: string, right: string) {
  return Buffer.compare(Buffer.from(left), Buffer.from(right));
}


function refKey(value: unknown): string | undefined {
  if (!object(value) || Object.keys(value).sort().join(",") !== "id,revision") return undefined;
  if (typeof value.id !== "string" || !UUID.test(value.id) || !Number.isSafeInteger(value.revision) || value.revision < 1) return undefined;
  return `${value.id}\u0000${String(value.revision).padStart(16, "0")}`;
}


function sortedUnique(values: unknown, key: (value: unknown) => string | undefined) {
  if (!Array.isArray(values)) return false;
  const keys = values.map(key);
  if (keys.some((value) => value === undefined)) return false;
  const concrete = keys as string[];
  const sorted = [...concrete].sort(compareBytes);
  return new Set(concrete).size === concrete.length && concrete.every((value, index) => value === sorted[index]);
}


function bindingKey(value: unknown) {
  if (!object(value) || typeof value.role !== "string") return undefined;
  const reference = refKey(value.node_ref);
  return reference === undefined ? undefined : `${value.role}\u0000${reference}`;
}


function namedKey(field: string) {
  return (value: unknown) => object(value) && typeof value[field] === "string" ? value[field] : undefined;
}


function collectRefs(unit: JsonObject) {
  const values: unknown[] = [];
  if (object(unit.concept_boundary)) values.push(...(unit.concept_boundary.focus_node_refs ?? []));
  for (const objective of unit.learning_objectives ?? []) if (object(objective)) values.push(...(objective.target_node_refs ?? []));
  for (const binding of unit.node_bindings ?? []) if (object(binding)) values.push(binding.node_ref);
  for (const behavior of unit.behaviors ?? []) if (object(behavior)) {
    values.push(...(behavior.input_node_refs ?? []));
    values.push(...(behavior.output_node_refs ?? []));
  }
  for (const validation of unit.validations ?? []) if (object(validation)) {
    values.push(...(validation.subject_node_refs ?? []));
    values.push(...(validation.evidence_node_refs ?? []));
  }
  for (const criterion of unit.mastery_criteria ?? []) if (object(criterion)) values.push(...(criterion.evidence_node_refs ?? []));
  return values.map(refKey).filter((value): value is string => value !== undefined);
}


function readStrict(path: string) {
  return strictParse(readFileSync(path));
}


export function validateUnit(unit: unknown, availableNodeRefs: unknown, contractRoot: string) {
  if (!object(unit)) return result(false, issue("knowledge_unit.invalid_json", ""));
  const missing = REQUIRED.find((field) => !(field in unit));
  if (missing !== undefined) return result(false, issue("knowledge_unit.missing_field", `/${missing}`));
  if (typeof unit.contract_version !== "string" || !unit.contract_version.startsWith("1.")) return result(false, issue("knowledge_unit.unsupported_contract_version", "/contract_version"));
  if (typeof unit.id !== "string" || !UUID.test(unit.id)) return result(false, issue("knowledge_unit.invalid_id", "/id"));
  if (!Number.isSafeInteger(unit.revision) || unit.revision < 1) return result(false, issue("knowledge_unit.invalid_revision", "/revision"));
  if (!sortedUnique(unit.node_bindings, bindingKey)) return result(false, issue("knowledge_unit.noncanonical_set", "/node_bindings"));
  if (!sortedUnique(unit.prerequisite_unit_refs, refKey)) return result(false, issue("knowledge_unit.noncanonical_set", "/prerequisite_unit_refs"));
  if (!sortedUnique(unit.provenance_refs, (value) => typeof value === "string" && value.length > 0 ? value : undefined)) return result(false, issue("knowledge_unit.noncanonical_set", "/provenance_refs"));
  for (const [field, identifier] of [["learning_objectives", "objective_id"], ["behaviors", "behavior_id"], ["validations", "validation_id"], ["mastery_criteria", "criterion_id"]]) {
    if (!sortedUnique(unit[field], namedKey(identifier))) return result(false, issue("knowledge_unit.noncanonical_set", `/${field}`));
  }
  const identity = refKey({ id: unit.id, revision: unit.revision });
  for (let index = 0; index < unit.prerequisite_unit_refs.length; index += 1) {
    if (refKey(unit.prerequisite_unit_refs[index]) === identity) return result(false, issue("knowledge_unit.self_dependency", `/prerequisite_unit_refs/${index}`));
  }
  for (let index = 0; index < unit.mastery_criteria.length; index += 1) {
    const criterion = unit.mastery_criteria[index];
    if (!object(criterion) || !Array.isArray(criterion.evidence_node_refs) || criterion.evidence_node_refs.length === 0) return result(false, issue("knowledge_unit.mastery_without_evidence", `/mastery_criteria/${index}/evidence_node_refs`));
  }
  for (let index = 0; index < unit.behaviors.length; index += 1) {
    const behavior = unit.behaviors[index];
    if (!object(behavior) || !CAPABILITIES.has(behavior.capability)) return result(false, issue("knowledge_unit.invalid_behavior_capability", `/behaviors/${index}/capability`));
  }
  if (!sortedUnique(availableNodeRefs, refKey)) return result(false, issue("knowledge_unit.noncanonical_set", "/available_node_refs"));
  const available = new Set((availableNodeRefs as unknown[]).map(refKey));
  const refs = collectRefs(unit);
  if (refs.some((reference) => !available.has(reference))) return result(false, issue("knowledge_unit.dangling_node_ref", ""));
  const bound = new Set(unit.node_bindings.map((binding: JsonObject) => refKey(binding.node_ref)));
  if (refs.some((reference) => !bound.has(reference))) return result(false, issue("knowledge_unit.dangling_node_ref", "/node_bindings"));
  const nestedSets: Array<[string, unknown]> = [];
  if (object(unit.concept_boundary)) {
    nestedSets.push(["/concept_boundary/focus_node_refs", unit.concept_boundary.focus_node_refs]);
    if (!sortedUnique(unit.concept_boundary.out_of_scope_statements, (value) => typeof value === "string" && value.length > 0 ? value : undefined)) return result(false, issue("knowledge_unit.noncanonical_set", "/concept_boundary/out_of_scope_statements"));
  }
  unit.learning_objectives.forEach((value: JsonObject, index: number) => nestedSets.push([`/learning_objectives/${index}/target_node_refs`, value.target_node_refs]));
  unit.behaviors.forEach((value: JsonObject, index: number) => {
    nestedSets.push([`/behaviors/${index}/input_node_refs`, value.input_node_refs]);
    nestedSets.push([`/behaviors/${index}/output_node_refs`, value.output_node_refs]);
  });
  unit.validations.forEach((value: JsonObject, index: number) => {
    nestedSets.push([`/validations/${index}/subject_node_refs`, value.subject_node_refs]);
    nestedSets.push([`/validations/${index}/evidence_node_refs`, value.evidence_node_refs]);
  });
  unit.mastery_criteria.forEach((value: JsonObject, index: number) => nestedSets.push([`/mastery_criteria/${index}/evidence_node_refs`, value.evidence_node_refs]));
  for (const [path, references] of nestedSets) {
    if (!sortedUnique(references, refKey)) return result(false, issue("knowledge_unit.noncanonical_set", path));
  }
  const schema = readStrict(resolve(contractRoot, "schemas/knowledge-unit.schema.json")) as JsonObject;
  if (!validateMachineSchema(unit, schema, schema, new Map())) return result(false, issue("knowledge_unit.invalid_json", ""));
  if (Buffer.byteLength(canonicalize(unit)) > KNOWLEDGE_UNIT_JCS_BYTES) return resourceExhausted();
  return result(true);
}


export function parseAndValidate(raw: Uint8Array, availableNodeRefs: unknown, contractRoot: string) {
  try {
    return validateUnit(strictParse(raw), availableNodeRefs, contractRoot);
  } catch (error) {
    if (!(error instanceof StrictJsonError)) throw error;
    return result(false, issue("knowledge_unit.invalid_json", ""));
  }
}


function walk(document: any, pointer: string): [any, string | number] {
  const parts = pointer.slice(1).split("/").map((part) => part.replaceAll("~1", "/").replaceAll("~0", "~"));
  let current = document;
  for (const part of parts.slice(0, -1)) current = Array.isArray(current) ? current[Number(part)] : current[part];
  return [current, Array.isArray(current) ? Number(parts.at(-1)) : parts.at(-1)!];
}


function mutate(unit: JsonObject, mutation: JsonObject) {
  const value = structuredClone(unit);
  const [parent, leaf] = walk(value, mutation.path);
  if (mutation.kind === "remove") delete parent[leaf];
  else if (mutation.kind === "reverse") parent[leaf] = [...parent[leaf]].reverse();
  else if (mutation.kind === "clear") parent[leaf] = [];
  else if (mutation.kind === "replace") parent[leaf] = structuredClone(mutation.value);
  else if (mutation.kind === "append-self-dependency") parent[leaf].push({ id: value.id, revision: value.revision });
  else throw new Error(`unsupported fixture mutation: ${mutation.kind}`);
  return value;
}


export function runFixtureSuite(contractRoot: string) {
  const suite = readStrict(resolve(contractRoot, "fixtures/cases.json")) as JsonObject;
  const base = suite.cases.find((fixture: JsonObject) => "unit" in fixture.input);
  const available = base.input.available_node_refs;
  const rows = suite.cases.map((fixture: JsonObject) => {
    const unit = "unit" in fixture.input ? structuredClone(fixture.input.unit) : mutate(base.input.unit, fixture.input.mutation);
    return { case_id: fixture.case_id, ...validateUnit(unit, available, contractRoot) };
  });
  return rows.sort((left: JsonObject, right: JsonObject) => compareBytes(left.case_id, right.case_id));
}
